libraries/AP_HAL_Linux/UARTDriver.cpp

```diff
@@ -553,22 +450,16 @@ void UARTDriver::_timer_tick(void)
     }
 
     // try to fill the read buffer
-    uint16_t _head;
-    n = BUF_SPACE(_readbuf);
-    if (n > 0) {
-        uint16_t n1 = _readbuf_size - _readbuf_tail;
-        if (n1 >= n) {
-            // one read will do
-            assert(_readbuf_tail+n <= _readbuf_size);
-            _read_fd(&_readbuf[_readbuf_tail], n);
-        } else {
-            assert(_readbuf_tail+n1 <= _readbuf_size);
-            int ret = _read_fd(&_readbuf[_readbuf_tail], n1);
-            if (ret == n1 && n > n1) {
-                assert(_readbuf_tail+(n-n1) <= _readbuf_size);
-                _read_fd(&_readbuf[_readbuf_tail], n - n1);
-            }
-        }
+    int ret;
+    ByteBuffer::IoVec vec[2];
+
+    const auto n_vec = _readbuf.reserve(vec, _readbuf.space());
+    for (int i = 0; i < n_vec; i++) {
+        ret = _read_fd(vec[i].data, vec[i].len);
+        assert(ret >= 0);
```

Comment:
You are right about the assert() purpose, I tried to replicate the totally missed the point, so +1 for removing it.

"Returning 0 here is completely fine... you should just break the loop here and stop reading."
You mean in _read_fd or something else? Because 0 is properly handled, the problem is negative values because of errors returned by _read_fd. So in the end this assert isn't that wrong and has a purpose, the consequences of quitting because an error code that is probably too extreme. Just break before commit should be enough, right ?